hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/ssl/ReloadingX509KeystoreManager.java

```diff
@@ -0,0 +1,151 @@
+
+    private static final Logger LOG = LoggerFactory.getLogger(ReloadingX509TrustManager.class);
+
+    static final String RELOAD_ERROR_MESSAGE =
+            "Could not load keystore (keep using existing one) : ";
+
+    private String type;
+    private String storePassword;
+    private String keyPassword;
+    private AtomicReference<X509ExtendedKeyManager> keyManagerRef;
+
+    /**
+     * Construct a <code>Reloading509KeystoreManager</code>
+     *
+     * @param type type of keystore file, typically 'jks'.
+     * @param location local path to the keystore file.
+     * @param storePassword password of the keystore file.
+     * @param keyPassword The password of the key.
+     * @throws IOException
+     * @throws GeneralSecurityException
+     */
+    public ReloadingX509KeystoreManager(String type, String location,
+                                        String storePassword, String keyPassword)
+            throws IOException, GeneralSecurityException {
+        this.type = type;
+        this.storePassword = storePassword;
+        this.keyPassword = keyPassword;
+        keyManagerRef = new AtomicReference<X509ExtendedKeyManager>();
+        keyManagerRef.set(loadKeyManager(Paths.get(location)));
+    }
+
+    @Override
+    public String chooseEngineClientAlias(String[] strings, Principal[] principals, SSLEngine sslEngine) {
+        return keyManagerRef.get().chooseEngineClientAlias(strings, principals, sslEngine);
+    }
+
+    @Override
+    public String chooseEngineServerAlias(String s, Principal[] principals, SSLEngine sslEngine) {
+        return keyManagerRef.get().chooseEngineServerAlias(s, principals, sslEngine);
+    }
+
+    @Override
+    public String[] getClientAliases(String s, Principal[] principals) {
+        return keyManagerRef.get().getClientAliases(s, principals);
+    }
+
+    @Override
+    public String chooseClientAlias(String[] strings, Principal[] principals, Socket socket) {
+        return keyManagerRef.get().chooseClientAlias(strings, principals, socket);
+    }
+
+    @Override
+    public String[] getServerAliases(String s, Principal[] principals) {
+        return keyManagerRef.get().getServerAliases(s, principals);
+    }
+
+    @Override
+    public String chooseServerAlias(String s, Principal[] principals, Socket socket) {
+        return keyManagerRef.get().chooseServerAlias(s, principals, socket);
+    }
+
+    @Override
+    public X509Certificate[] getCertificateChain(String s) {
+        return keyManagerRef.get().getCertificateChain(s);
+    }
+
+    @Override
+    public PrivateKey getPrivateKey(String s) {
+        return keyManagerRef.get().getPrivateKey(s);
+    }
+
+    public ReloadingX509KeystoreManager loadFrom(Path path) {
+        try {
+            this.keyManagerRef.set(loadKeyManager(path));
+        } catch (Exception ex) {
+            // The Consumer.accept interface forces us to convert to unchecked
+            throw new RuntimeException(ex);
```

Comment:
What will happen when this comes out? Where will it be caught? Will it cause damage causing process exit or thread exit?

Reply:
It gets caught by the FileMonitoringTimerTask and from there it's logged. It won't cause any damage, but it will keep logging the error on every reload interval if it's not resolved, potentially filling the logs and disk space.